### PR TITLE
Stop promoting all the things

### LINKED ISSUE #91009: Values that need to be dropped are promoted in const / static initializers
For example, promotion succeeds in the example below, although compilation nevertheless fails due to checks for live drops:

```rust
pub const fn id<T>(x: T) -> T { x }
pub const C: () = {
    let _: &'static _ = &String::new();
    // Promotion failed, two errors:
    // ERROR: destructors cannot be evaluated at compile-time
    // ERROR: temporary value dropped while borrowed

    let _: &'static _ = &id(&String::new());
    // Promoted. Only one error:
    // ERROR: destructors cannot be evaluated at compile-time

    let _: &'static _ = &std::mem::ManuallyDrop::new(String::new());
    // Promoted. No errors.
};
```

With `const_precise_live_drops`, checking for live drops is delayed until after the promotion, so the following example compiles successfully:

```rust
#![feature(const_precise_live_drops)]
pub const fn id<T>(x: T) -> T { x }
pub const C: () = {
    let _: &'static _ = &id(&String::new());
};
```

The check for live drops can also be passed by providing a const drop implementation. The following example compiles successfully as well:

```rust
#![feature(const_fn_trait_bound)]
#![feature(const_mut_refs)]
#![feature(const_trait_impl)]
struct Panic;
impl const Drop for Panic { fn drop(&mut self) { panic!(); } }
pub const fn id<T>(x: T) -> T { x }
pub const C: () = {
    let _: &'static _ = &id(&Panic);
};
```

@rustbot label +A-const-eval

### PR DESCRIPTION
**Pull request by oli-obk**
fixes #91009

r? @RalfJung

### CODE REVIEW

**Comment by oli-obk** [comment]
@craterbot check

**Comment by RalfJung** [comment]
That is not the same test. Note that the one I linked did a bunch of stuff to register functions as `rustc_promotable` which is important.

File: src/test/ui/consts/promoted_const_call.rs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    let _: &'static _ = &id(&Panic);
+    //~^ ERROR: temporary value dropped while borrowed
+    //~| ERROR: temporary value dropped while borrowed
+};
+
+fn main() {
+    let _: &'static _ = &id(&Panic);
+    //~^ ERROR: temporary value dropped while borrowed
+    //~| ERROR: temporary value dropped while borrowed
+}
```
Comment (RalfJung) [inline src/test/ui/consts/promoted_const_call.rs:16]: Do we need these const tests to be 3 separate tests or can they be one test?

**Comment by RalfJung** [comment]
And [here is another testcase](https://github.com/rust-lang/rust/issues/91009#issuecomment-1333772650) :)

**Comment by RalfJung** [comment]
Actually we can make that a run-pass test instead -- might be good to have one like that as well:
```rust
use std::sync::atomic::*;

static FLAG: AtomicBool = AtomicBool::new(false);

struct NoisyDrop(&'static str);
impl Drop for NoisyDrop {
    fn drop(&mut self) {
        FLAG.store(true, Ordering::SeqCst);
    }
}
fn main() {
    {
        let _: &'static _ = &&(NoisyDrop("drop!"), 0).1;
    }
    assert!(FLAG.load(Ordering::SeqCst));
}
```

File: src/test/ui/consts/promoted_const_call.rs (hunk 1 above)
Comment (oli-obk) [reply src/test/ui/consts/promoted_const_call.rs:19]: Each of these uses different feature gates (or none)

**Comment by RalfJung** [comment]
r=me on code and tests, let's see what crater says.

Will this need t-lang FCP?

**Comment by oli-obk** [comment]
The only regression is https://github.com/slightlyoutofphase/staticvec/blob/64c0a85c04b2ffc27ea3b20b87097f0dc49968f3/src/lib.rs#L31 which uses so many unstable features I think we can safely break them. cc @slightlyoutofphase

**Comment by oli-obk** [comment]
Nominating for T-lang:

This PR is a breaking change fixing a hole in our promotion checks. Most notably:

```rust
struct NoisyDrop(&'static str);
impl Drop for NoisyDrop {
    fn drop(&mut self) {
        println!("{}", self.0);
    }
}
fn main() {
    let _ = &&(NoisyDrop("drop!"), 0).1;
    // NosiyDrop is not dropped ...
}
```

does not drop the `NoisyDrop`, as it gets implicitly promoted. This is obviously a bug, as the user did not ask for promotion and we should not silently promote away side effects like dropping.

Crater has found a single breaking change in a nightly-only crate.

@rust-lang/wg-const-eval would like to merge this PR.

**Comment by joshtriplett** [comment]
Discussed in today's @rust-lang/lang meeting. Looks good, let's confirm consensus via FCP:

@rfcbot merge

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

**Comment by pnkfelix** [comment]
* The three primary regressions are all to unicode-normalization {check-full, debug-full, check-incr-full}, by 1% or less.
* The secondary regressions are to coercions-debug-full (by 4%), and to variations of ucd (by 1.25% or less).
* In my judgement, the semantic problem this addresses more than offsets the cost that this PR is paying. Marking as triaged.

@rustbot label: +perf-regression-triaged

**Pull request merged**